File: src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoRemount.java
package me.zeroeightsix.kami.module.modules.movement;

import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityBoat;
import net.minecraft.entity.passive.*;
import net.minecraft.util.EnumHand;

import java.util.Comparator;

/**
 * @author dominikaaaa
 * Created by dominikaaaa on 05/04/20
 * updated by ionar2 on 04/05/20
 * Updated by dominikaaaa on 05/06/20
 */
@Module.Info(
        name = "AutoRemount",
        description = "Automatically remounts your horse",
        category = Module.Category.MOVEMENT
)
public class AutoRemount extends Module {
    private Setting<Boolean> boat = register(Settings.b("Boats", true));
    private Setting<Boolean> horse = register(Settings.b("Horse", true));
    private Setting<Boolean> skeletonHorse = register(Settings.b("SkeletonHorse", true));
    private Setting<Boolean> donkey = register(Settings.b("Donkey", true));
    private Setting<Boolean> mule = register(Settings.b("Mule", true));
    private Setting<Boolean> pig = register(Settings.b("Pig", true));
    private Setting<Boolean> llama = register(Settings.b("Llama", true));
    private Setting<Float> range = register(Settings.floatBuilder("Range").withMinimum(1.0f).withValue(2.0f).withMaximum(10.0f).build());

    public void onUpdate() {
        // we don't need to do anything if we're already riding.
        if (mc.player.isRiding())
            return;

        mc.world.loadedEntityList.stream()
                .filter(this::isValidEntity)
                .min(Comparator.comparing(en -> mc.player.getDistance(en)))
                .ifPresent(entity -> mc.playerController.interactWithEntity(mc.player, entity, EnumHand.MAIN_HAND));

    }

    private boolean isValidEntity(Entity entity) {
        if (mc.player.isRiding()) return false;

        if (entity.getDistance(mc.player) > range.getValue()) return false;

        if (entity instanceof AbstractHorse) {
            AbstractHorse horse = (AbstractHorse) entity;

            /// no animal abuse done in this module, no thanks.
            if (horse.isChild()) return false;
        }

        if (entity instanceof EntityBoat && boat.getValue()) return true;

        if (entity instanceof EntityHorse && horse.getValue()) return true;

        if (entity instanceof EntitySkeletonHorse && skeletonHorse.getValue()) return true;

        if (entity instanceof EntityDonkey && donkey.getValue()) return true;

        if (entity instanceof EntityMule && mule.getValue()) return true;

        if (entity instanceof EntityPig && pig.getValue()) {
            EntityPig pig = (EntityPig) entity;
            return pig.getSaddled();
        }

        return entity instanceof EntityLlama && llama.getValue();
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/movement/AutoRemount.kt
package me.zeroeightsix.kami.module.modules.movement

import me.zeroeightsix.kami.event.SafeClientEvent
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import me.zeroeightsix.kami.util.TickTimer
import me.zeroeightsix.kami.util.TimeUnit
import me.zeroeightsix.kami.util.threads.safeListener
import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityBoat
import net.minecraft.entity.passive.*
import net.minecraft.util.EnumHand
import net.minecraftforge.fml.common.gameevent.TickEvent

object AutoRemount : Module(
    name = "AutoRemount",
    description = "Automatically remounts your horse",
    category = Category.MOVEMENT
) {
    private val boat = setting("Boats", true)
    private val horse = setting("Horse", true)
    private val skeletonHorse = setting("SkeletonHorse", true)
    private val donkey = setting("Donkey", true)
    private val mule = setting("Mule", true)
    private val pig = setting("Pig", true)
    private val llama = setting("Llama", true)
    private val range = setting("Range", 2.0f, 1.0f..5.0f, 0.5f)
    private val remountDelay = setting("RemountDelay", 5, 0..10, 1)

    private var remountTimer = TickTimer(TimeUnit.TICKS)

    init {
        safeListener<TickEvent.ClientTickEvent> {
            // we don't need to do anything if we're already riding.
            if (player.isRiding) {
                remountTimer.reset()
                return@safeListener
            }
            if (remountTimer.tick(remountDelay.value.toLong())) {
                world.loadedEntityList.stream()
                        .filter { entity: Entity -> isValidEntity(entity) }
                        .min(compareBy { player.getDistance(it) })
                        .ifPresent { playerController.interactWithEntity(player, it, EnumHand.MAIN_HAND) }
            }
        }
    }

    private fun SafeClientEvent.isValidEntity(entity: Entity): Boolean {
        if (entity.getDistance(player) > range.value) return false
        return entity is EntityBoat && boat.value
                || entity is EntityAnimal && !entity.isChild // FBI moment
                && (entity is EntityHorse && horse.value
                || entity is EntitySkeletonHorse && skeletonHorse.value
                || entity is EntityDonkey && donkey.value
                || entity is EntityMule && mule.value
                || entity is EntityPig && entity.saddled && pig.value
                || entity is EntityLlama && llama.value)
    }
}